Remove useless optimisation

istarkov opened the pull request:
Removed optimisation is useless in React 16. The only intent for it was to not update if null was returned in handler. 
But returning same value didn't caused update too, this is not same as React setState works.
This is possibly a breaking change.

istarkov commented:
See https://github.com/istarkov/rifm/issues/27#issuecomment-417255418

volkanunsal commented:
Hi! I started getting [this error](https://reactjs.org/docs/error-decoder.html/?invariant=185) when I upgraded to version 0.30.1. I think this PR might be the culprit. Do you know how I can fix this error?

istarkov commented:
Just check your code, what causes setState on render. Usually this happens because of incorrect usage of setState, inside lifecycle handlers like `componentDidUpdate`, `componetWillReceiveProps` etc. Also this optimisation did nothing in modern (latest versions) of React.

volkanunsal commented:
Hm, ok. I'm using React 16.4.0. I do use setState in `componentWillReceiveProps`, i.e. an updater from `withStateHandlers`, all over the place in my code. I didn't realize that was bad practice. Can you point me to where it says that?

Is there any harm in leaving this optimization in place given that some of us depend on its behavior?

istarkov commented:
> Is there any harm in leaving this optimization in place given that some of us depend on its behavior?
No without example where will be shown that exactly removal of this optimisation caused the issue and the example code correct.
As componentWillReceiveProps is deprecated in React. 
Also about bad practice - if you update state of the parent in cwrp depending on props it can cause infinite loop - its overall not good (but not so bad) practice to change state depending on props change. Here in issues there were few examples about. Same for cdu lifecycle having that it can cause loop changing his own state (not parent only)

volkanunsal commented:
> No without example where will be shown that exactly removal of this
optimisation caused the issue

This is the only major change since the last release. What else could it
be?

You’re correct that cwrp is being deprecated but it won’t be removed until
React version 17. So until then maybe it makes sense to not break the
existing apps with this optimization? We need time to upgrade.

Removing this feature is not super critical right now. It’s not doing
anyone any harm even if it’s encouraging a bad practice. Why not wait until
React 17 and release it as part of a major version upgrade?

On Tue, Oct 23, 2018 at 1:33 PM Ivan Starkov <notifications@github.com>
wrote:

> Is there any harm in leaving this optimization in place given that some of
> us depend on its behavior?
>
> No without example where will be shown that exactly removal of this
> optimisation caused the issue and the example code correct.
>
> As componentWillReceiveProps is deprecated in React.
> Also about bad practice - if you update state of the parent in cwrp
> depending on props it can cause infinite loop - its overall not good (but
> not so bad) practice to change state depending on props change. Here in
> issues there were few examples about. Same for cdu lifecycle having that it
> can cause loop changing its own state (not parent only)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/acdlite/recompose/pull/736#issuecomment-432343355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAJQMOVSfK8AQtIVgHiK9PHmxv3vNYJpks5un1LUgaJpZM4WTJGD>
> .
>
-- 
*Volkan Unsal*
*Product Engineer*
volkanunsal.com

istarkov commented:
> This is the only major change since the last release. What else could it
be?

I'm more than 20 years in development, be sure I know the answer on "What else could it be".
And the answer is - IT COULD BE ANYTHING ;-)

volkanunsal commented:
Come on now. You can't be serious.

volkanunsal commented:
I'll take my time to test it more carefully, and maybe you're right.
